blockchain/pallets/staking/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+            DispatchClass::Normal,
+            Pays::No
+        ))]
+        pub fn set_minimum_stake(
+            origin: OriginFor<T>,
+            min: Option<BalanceOf<T>>,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+
+            MinimumStake::<T>::set(min);
+
+            Ok(())
+        }
+
+        #[pallet::weight((
+            10_000 + T::DbWeight::get().reads_writes(0, 1),
+            DispatchClass::Normal,
+            Pays::No
+        ))]
+        pub fn set_lock_period_shares(
+            origin: OriginFor<T>,
+            #[pallet::compact] blocks: BlockNumberFor<T>,
+            #[pallet::compact] shares: u32,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+
+            LockPeriodShares::<T>::insert(blocks, shares);
+
+            Ok(())
+        }
+
+        #[pallet::weight(10_000 + T::DbWeight::get().reads_writes(4, 2))]
+        pub fn stake(
+            origin: OriginFor<T>,
+            #[pallet::compact] lock_period: BlockNumberFor<T>,
+            #[pallet::compact] amount: BalanceOf<T>,
+        ) -> DispatchResult {
+            let address = ensure_signed(origin)?;
+
+            if let Some(min) = MinimumStake::<T>::get() {
+                if amount < min {
+                    return Err(Error::<T>::AmountBelowMinimum.into());
+                }
+            }
+
+            let shares =
+                LockPeriodShares::<T>::get(lock_period).ok_or(Error::<T>::UnknownLockPeriod)?;
+
+            T::Currency::reserve(&address, amount)
+                .map_err(|_| Error::<T>::CannotStakeMoreThanBalance)?;
+
+            StakedAmounts::<T>::mutate(
+                address,
+                <frame_system::Pallet<T>>::block_number() + lock_period,
+                |map| {
+                    *map.entry(shares).or_insert_with(BalanceOf::<T>::zero) += amount;
+                },
+            );
+            TotalCoinShares::<T>::mutate(|b| *b += amount * shares.into());
+
+            Ok(())
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        pub fn staked_balance(account: T::AccountId) -> BalanceOf<T>
+        where
+            BalanceOf<T>: core::iter::Sum,
+        {
+            StakedAmounts::<T>::iter_prefix_values(account)
+                .map(|map| map.values().cloned().sum())
+                .sum()
+        }
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T>
+    where
+        BalanceOf<T>: core::iter::Sum,
+    {
+        fn on_finalize(block_number: BlockNumberFor<T>) {
+            if block_number % T::DistributeEveryNBlocks::get() != 0u32.into() {
+                return;
+            }
+
+            let to_distribute = T::Currency::free_balance(&T::DistributionSource::get());
+            let total_staked = TotalCoinShares::<T>::get();
+
+            let mut distributed = BalanceOf::<T>::zero();
+            let mut distributed_shares = BalanceOf::<T>::zero();
```

Comment:
put a comment on top explaining the formula

````
// The formula is roughly
// user_shares = % of total staked FCL + length of time
// earned_fcl = total_fcl_to_distribute * user_shares / total_shares
```

Reply:
Refactored a lot. Hopefully the logic is clearer now.